### PR TITLE
Integration of video and audio into the metadata editor 

### LINKED ISSUE #4248: Image frames without thumbnails not visible in the metadata editor
If there are images but no thumbnails, a lot of invisible image frames populate the gallery of the metadata editor. They only get visible if you move the mouse over them.

**Goal:** The frames should be clearly visible and you should be able to work with them just as if there were images (as long you know what you are doing)

### PR DESCRIPTION
**Pull request by markusweigelt**
- Extend media component for streaming videos
- Add streaming and buffering handler for videos
- Refactor gallery with partials of list elements and detail page
- Extend the project configuration to select media variants for preview and media view of video gallery items

There already existed a PR #4870 in draft status was quite ready. Because the release management has changed, draft PR's where closed. The implementation from the old PR is used and finalized in this PR.

Small documentation on how to configure videos has already been deposited in the wiki.
https://github.com/kitodo/kitodo-production/wiki/Video-f%C3%BCr-den-Metadaten-Editor-konfigurieren

Fixes #4248

### CODE REVIEW

**Review by matthias-ronge — COMMENTED** [review]
Overall, great thing that you make video available in PP!

Overall, great thing that you make video available in PP!

The only change that I think is necessary here is rename the enum TYPE, and put it in a separate class file. The rest is mainly thought out loud, so I leave it up to you what you want to take from it.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java
Comment (markusweigelt) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java:130]: It is not possible to say exactly at the moment. For example, the preview images can be pre-generated from the video and so you can decide which media type variant you want to use as a preview. Besides the thumbnails can be moved to a separate folder, which may correspond to the aspect ratio of the movie. I would keep open to create preview video, to use seperate generated thumbnails or just use the existing video as a preview.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java
Comment (markusweigelt) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java:137]: A use case would be that you don't want to work directly with the videos for metadata acquistion, but only with their generated thumbnails. `effective webwork` had such a realization in a project with video types.

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryPanel.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryPanel.java:521]: To concatenate a string, you need to declare a string and use the + operator and Java do the rest. No explizit valueOf is needed. For the Integer type the toString() method is called behind. 
https://www.quora.com/Does-Java-allow-strings-and-integers-to-be-concatenated

File: Kitodo/src/main/java/org/kitodo/production/forms/ProjectForm.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/forms/ProjectForm.java:546]: Good note! 👍🏻 
All necessary video mime types are starting with `video` https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types

For HTML Video tag only following formats are possible https://www.w3schools.com/tags/tag_video.asp. 

So if there is a video format with `application/x-`, the browser will at least not support it.

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java:240]: PrimeResourceHandler should close the input stream.

https://searchcode.com/file/2686530/src/main/java/org/primefaces/application/PrimeResourceHandler.java/

File: Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java]: Done :)

File: Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css
Comment (markusweigelt) [reply Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css:3038]: It is a common way to use such selector to style by attribute values. If there is a convention or this is not wanted by the majority then I take it out. The advantage here is that no `IF ELSE` has to be used in the template to render the class.

Comment (oliver-stoehr) [reply Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css:3039]: As far as I know we don't have any conventions regarding this. I think it's fine to use data attributes in CSS selectors, especially when a CSS class would be a redundant information.

Comment (solth) [reply Kitodo/src/main/webapp/WEB-INF/resources/css/kitodo.css:3039]: According to Codacy the regex ( `data-mime-type^="video"`) in the CSS selector can negatively impact performance and should be avoided (see Codacy report of current build for details). 

Since this code is part of the metadata editor, we should probably be extra vigilant and strive not to loose the performance improvements we gained in the last year. So perhaps a CSS class would be preferable after all?

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java:36]: Please add some java doc to this class and it's methods to explain what it does.

File: Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java:64]:
```suggestion
        if (controlsParameter.isEmpty()  || !Boolean.FALSE.toString()
```

File: Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java
Comment (solth) [inline Kitodo/src/main/java/org/kitodo/production/renderer/KitodoMediaRenderer.java:101]: This method seems to be never used. Do we really need it? Otherweise, please add it once it is required.

**Review by solth — CHANGES_REQUESTED** [review]
@markusweigelt thanks a lot for this pull request! This seems to be a great new feature for the metadata editor.

This is just a first pass of the review with a few remarks. I will check the xhtml files and test the functionality separately. 

One important thing I can already say: For new functionality and changes of this scope, tests are required!

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java:241]: Sorry may comment was not right and long time since i implemented the RangeStreamContentHandler. I have adapted the primeresourcehandler like this. The input stream is also closed in the RangeStreamContentHandler. https://github.com/kitodo/kitodo-production/blob/39337a5949e5d03e888f67a6368f11ef0419424e/Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java#L90

Nevertheless, I fit the try block.

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java
Comment (matthias-ronge) [inline Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java:162]: “Return 1 part” reads like “Return part nr. one”. Maybe change to “returning part of file”

**Review by matthias-ronge — CHANGES_REQUESTED** [review]
Thank you for this very powerful development! This will take us to a new level!

I would like two changes:

1.) You need a distinction between image, video and sound. Please implement this in fileFormats.config and don't invent anything new here. We already have a configuration option for file types. That should go here. I'm happy to support you! We can change that together.

2.) Please review your PR for changes that only change the layout of the code and revert them. This makes it much easier for the reviewer.

**Comment by markusweigelt** [comment]
> 1.) You need a distinction between image, video and sound. Please implement this in fileFormats.config and don't invent anything new here. We already have a configuration option for file types. That should go here. I'm happy to support you! We can change that together.

I have already adjust `fileFormats.config` (see in linked documentation https://github.com/kitodo/kitodo-production/wiki/Video-f%C3%BCr-den-Metadaten-Editor-konfigurieren) with video type. Otherwise, the folders from which the video files are listed cannot be created. 

In our case we need different types of media types, e.g. for the structuring of an audio carrier, we have besides the actual media file also images (e.g. Record page, ...) which should be assigned to the respective track. That means we have several types of media files (multiple mediavariants) in the list view and not one. 

image

image

For videos the old variant would work for us too but I made it configurable for videos as well to audio (atm not in documentation) because we really need it for audio and it's just more customizable e.g. to configure scene images for video tracks. (Atm it is a draft PR but it works https://github.com/kitodo/kitodo-production/pull/5504)

That's why I wanted to make the views configurable for the respective type Video and Audio to current config as well. It is no longer sufficient to select only one type from the variants.
So maybe it's even better to solve this more generically here. In my opinion it would make sense for each MediaContentType to make the media variant configurable for the preview and also the detail.

> 2.) Please review your PR for changes that only change the layout of the code and revert them. This makes it much easier for the reviewer. 

Yes that is absolutely correct. I configured to format only changed lines but sometimes i add code that is then removed again because i found a better way and the line is formatted. But I will pay more attention to it! In my case i use the https://github.com/kitodo/kitodo-production/blob/master/config/Kitodo-IDE-formatting-Eclipse.xml with the formatting when saving. The formatting seems to have changed as well or there is a lot code with old formatting, otherwise the formatting of the source code would not be adjusted.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java
Comment (markusweigelt) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java:130]: I comment this here as answer to your top level comment. https://github.com/kitodo/kitodo-production/pull/5472#issuecomment-1381985218

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java
Comment (markusweigelt) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Project.java:137]: I comment this here as answer to your top level comment. https://github.com/kitodo/kitodo-production/pull/5472#issuecomment-1381985218

File: Kitodo/pom.xml
Comment (markusweigelt) [reply Kitodo/pom.xml:126]: The `javax.el` is already used in Kitodo-DataManagement and available dependency management of main pom.xml. I need this for `javax.el.ValueExpression` import in `RangeStreamedContentHandler`. This handler is an extended implementation of `StreamedContentHandler` and `PrimefaceResourceHandler` also use the class.

File: Kitodo/src/main/java/org/kitodo/production/enums/MediaContentType.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/enums/MediaContentType.java:15]: Good idea I would also implement this immediately in the XML but this is not what I need. I need the base type of the media variant. 

The media files of PhysicalDivision consists of several mime types. In the current implementation I have implemented it so: if the media files contains a file which has a mimetype video, than the MediaVariant of the type video.  https://github.com/kitodo/kitodo-production/blob/008c656b5df7cbd2d8347f3f50a226458e7f81fd/Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryPanel.java#L525

**Comment by matthias-ronge** [comment]
I now understand the problem.

For audio you really need two full view folders, one with the audio and one with the picture. If it is mixed in the process, that there is audio with still image, and video, then you need even three (because not every folder will have content for each `MediaUnit`).

In that case I would:
- rename `mediaView` to `picture`
- (maybe: rename `preview` to `thumbnail`?)
- add `audio` and `video`

In my mind, there should only be thumbnails in the gallery, no video or sound, otherwise the gallery becomes too confusing, I think. Or is this a _must_ requirement?

If it is absolutely required, I would use the same `sound` and `video` content in the gallery, and leave the downscaling of videos to the browser. I think nobody will in practice generate downscaled video clips just for use in the Production gallery, because it would take a lot of resources to generate them. The reason for the thumbnails is, that they have to load all at the same moment when showing the page, while typically only one video should start playing in the gallery, when the mouse moves over it. Or am I wrong?

File: Kitodo/src/main/java/org/kitodo/production/enums/MediaContentType.java
Comment (matthias-ronge) [reply Kitodo/src/main/java/org/kitodo/production/enums/MediaContentType.java:15]: I see that this doesn’t solve your problem, but however classifying the media types as `picture`, `sound`, `video` and `other` would make sense (also for populating the drop-downs only with folders that make sense in them). But we can do this in a later pull request.

File: Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/GalleryMediaContent.java]: The following line `InputStream viewData = ServiceManager.getFileService().read(uri)` in try brackets, does not work for me. The video returns a server error ("input stream already closed") and is not rendered in the HTML video tag. Probably the stream is closed too early, so the `RangeStreamContentHandler` can't handle this stream. So I think that closing the stream only in the `RangeStreamContentHandler`, as described in the comment above, will be sufficient.

**Comment by markusweigelt** [comment]
I have added tests and made the media variants for audio selectable in project. If no media variant is selected, icons are displayed for type image, video and audio.

Cause of https://github.com/kitodo/kitodo-production/issues/5515 I will not rename "mediaview" and "preview" for now.

For the test, I did a refactoring, removing PowerMock, which is no longer being developed. Mockito has been brought to the latest version.

**Comment by solth** [comment]
Does this solve #4248, too?

**Comment by markusweigelt** [comment]
Yes when no folder is selected for the list view under settings of project, icons are displayed now.

**Comment by solth** [comment]
Works very well, just two small quirks I encounterd so far:
- when selecting the video folder as preview for videos, the video player controls are displayed in the thumbnails, even though the videos cannot be started there; here, the controls should be hidden:
Bildschirm­foto 2023-02-01 um 14 27 18
- once configured, I cannot select the "Deactivated" option for the folders for "Video media view" and "Video media preview" in the project settings without first selecting another, intermediate folder before selecting the "Deactivated" option:
https://user-images.githubusercontent.com/19183925/216057090-17e13ccd-4ff8-407b-975b-9a9047ecbc9e.mp4

**Review by solth — COMMENTED** [review]
No major problems found, just some minor comments.

Thumbs up for moving some parts of the gallery to external xhtml files that are added via `include` tags!

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java:270]: Change visibility to private cause there is no need for public visibility.

File: Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/partials/media-list.xhtml
Comment (markusweigelt) [reply Kitodo/src/main/webapp/WEB-INF/templates/includes/metadataEditor/partials/media-list.xhtml:44]: Native loading attribute is only available for images and iframe https://caniuse.com/loading-lazy-attr. In Firefox only for images. Lazy loading can be added for videos and audio using appropriate javascript libraries but not native. 
I assume for now that within the gallery only a few videos and audios are displayed, which are visible within the display area. If there is a need in the future to load more videos we can add such a lib here.

File: Kitodo/src/test/java/org/kitodo/BaseMockitoTest.java
Comment (markusweigelt) [reply Kitodo/src/test/java/org/kitodo/BaseMockitoTest.java:19]: Yes i reduce this code cause there was a advantage but i changed implementation. So i will remove the base class and add `@RunWith` to inherited class.

File: Kitodo/src/test/java/org/kitodo/production/handler/RangeStreamedContentHandlerIT.java
Comment (markusweigelt) [reply Kitodo/src/test/java/org/kitodo/production/handler/RangeStreamedContentHandlerIT.java]: There is no reason for this. The code formatter doesn't seem to take this into account and I haven't found a setting for it in IntelliJ code formatter settings. If you have a hint for this I would be very happy! I have added the whitespaces manually now.

**Comment by markusweigelt** [comment]
> * when selecting the video folder as preview for videos, the video player controls are displayed in the thumbnails, even though the videos cannot be started there; here, the controls should be hidden:
> 

Solved.

> * once configured, I cannot select the "Deactivated" option for the folders for "Video media view" and "Video media preview" in the project settings without first selecting another, intermediate folder before selecting the "Deactivated" option:
> 
>  Bildschirmaufnahme.2023-02-01.um.14.30.25.mp4

The problem here is that no METS filegroup was selected for the folder. With selected METS filegroup it works. Probably it is because of the comma at the end for the Value which the selectOneMenu component of Primeface doesn't seem to handle. 

You could solve the problem in different ways:

One variant would be for METS group to make the first option selectable as default when adding the folder. Currently once a METS filegroup is selected it cannot be deselected. (Most sustainable and simplest variant)

Second variant would be to use the field as mandatory field.

Variant three find the cause of the error in the component.

File: Kitodo-DataManagement/src/main/resources/db/migration/V2_121__Add_selectable_audio_video_preview_and_mediaview_folders_to_project.sql
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/resources/db/migration/V2_121__Add_selectable_audio_video_preview_and_mediaview_folders_to_project.sql:17]: Should it not be `FK_preview_video_folder_id` instead of current `CK_preview_video_folder_id`  where `FK` is standing for "foreign key"?  If this `FK` is correct then it should be changed in the whole file.

**Comment by solth** [comment]
@markusweigelt thanks a lot for implementing the requested changes. From my perspective, the PR is approved, but since @matthias-ronge and @oliver-stoehr have more experience with the concept and implementation of file groups in the project settings I would like to wait for their review before merging the pull request.

**Review by solth — COMMENTED** [review]
@markusweigelt thanks a lot for implementing the requested changes. From my perspective, the PR is approved, but since @matthias-ronge and @oliver-stoehr have more experience with the concept and implementation of file groups in the project settings I would like to wait for their review before merging the pull request.

**Review by solth — APPROVED** [review]

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java
Comment (matthias-ronge) [inline Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java:53]: Though it qualifies as a constant, we use lowercase `logger` variable name all through the project

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java
Comment (matthias-ronge) [inline Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamedContentHandler.java:247]: Please put class in its own file

**Review by matthias-ronge — CHANGES_REQUESTED** [review]
I see in your HTML you sometimes use `fn:startsWith(field.previewMimeType, 'video')` and sometimes `field.type eq 'VIDEO'`. _If possible_ I would suggest to stick to the second one only.

**Review by matthias-ronge — APPROVED** [review]

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamContentHandler.java
Comment (oliver-stoehr) [inline Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamContentHandler.java:42]:
```suggestion
 * This handler allows to stream files e.g. video and audio partially to the browser. Only parts of the file that are
```

**Review by oliver-stoehr — CHANGES_REQUESTED** [review]
This is a great addition to the metadata editor!

It seems to work fine except one scenario:
I configured one file group for videos and set this for "media view for videos". But I set the "preview for videos" to "preview for videos disabled".
With this combination the fallback video icon is correctly shown in the structured view, but the video cannot be streamed in the detail view. The Ajax request for the stream is answered with an HTTP 500 (I could not find any error in the log files).

File: Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamContentHandler.java
Comment (oliver-stoehr) [reply Kitodo/src/main/java/org/kitodo/production/handler/RangeStreamContentHandler.java:42]: (This is just a minor typo not related to the described scenario.)

**Comment by markusweigelt** [comment]
@oliver-stoehr Thank you for the review! I will check this.

Does it work with the other combinations for you? Were you able to play videos and audios?

**Comment by oliver-stoehr** [comment]
Yes, the streaming in detail view works if I set a file group as preview.

**Comment by markusweigelt** [comment]
@oliver-stoehr I have fixed the error. It should also work in the variant with the disabled preview now.

**Review by oliver-stoehr — APPROVED** [review]

**Comment by solth** [comment]
@markusweigelt since all reviewers have approved the pull request now, could you do one last rebase please?

**Pull request merged**
